base/ltcmd.dtx

```diff
@@ -437,16 +437,52 @@
 %
 % \begin{macro}{\@@_declare_cmd_code:Nnn}
 % \begin{macro}
-%   {\@@_declare_cmd_code_aux:Nnn, \@@_declare_cmd_code_expandable:Nnn}
-%   The appropriate auxiliary is called.
+%   {
+%     \@@_declare_cmd_optimised:Nnn,
+%     \@@_declare_cmd_code_aux:Nnn,
+%     \@@_declare_cmd_code_expandable:Nnn
+%   }
+% \begin{macro}{\@@_start_optimised:}
+% \changes{v1.2b}{2023/11/28}
+%   {Optimise cmd creation for all-\texttt{m} arguments}
+%   At this stage we can check for a short-cut possibility: if the signature
+%   is just made up of \cs{@@_expandable_grab_m:w}, then we can produce an
+%   optimised document command. This only applies to document commands,
+%   not creation of environments (which are more complex).
 %    \begin{macrocode}
 \cs_new_protected:Npn \@@_declare_cmd_code:Nnn
   {
-    \bool_if:NTF \l_@@_grab_expandably_bool
-      { \@@_declare_cmd_code_expandable:Nnn }
-      { \@@_declare_cmd_code_aux:Nnn }
+    \bool_lazy_and:nnTF
+      { ! \l_@@_environment_bool }
+      {
+        \str_if_eq_p:ee
+          { \exp_not:V \l_@@_signature_tl }
+          {
+            \prg_replicate:nn \l_@@_current_arg_int
+              { \exp_not:N \@@_expandable_grab_m:w }
+          }
```

Comment:
Looks like this can be extended to cover all-`\@@_expandable_grab_m_long:w` signature too, which is closer to what `\newcommand` defines.

Reply:
Yes, did occur to me: I'll look at that once the first pass is done

Reply:
Addressed by 62d512955111a94a6b8dae3d183de66ddeb0a48c - should make us as efficient as `\def`, more or less, for the common case.